### ISSUE TITLE
Method Scanning

### ISSUE THREAD

**Issue by zapbot** [zaproxy/zaproxy#174]
```
In addition to portscanning, it would be nice if ZAP can perform method scanning (GET,
HEAD, POST, PUT, DELETE, etc) and flag the dangerous allowed methods.


```

Original issue reported on code.google.com by `houcem.hachicha` on 2011-09-18 11:05:05

**Comment by vanderaj** [comment]
Modern web applications require OPTIONS to exist, and in fact, when access control rules are present, browsers issue a pre-flight OPTIONS request to determine if the request is permitted and will in fact succeed. This is normal and expected behavior. 

https://developer.mozilla.org/en-US/docs/Web/HTTP/Access_control_CORS

Scanning for PUT and DELETE is hazardous; I would suggest breaking this into a WebDav check (which uploads or deletes a test file with a random filename), and NOT to use an active permutable scan against RESTful API where the current session context might destroy the application, such as PUTting a new configuration item or DELETING a critical admin context. You don't see it much these days, but a Frontpage / Sharepoint check for _vti* and author.dll should also happen. This is an oldie but well abused extensively in the past. This might be better off as yet another active scan check. 

I would also suggest testing for how the app handles HEAD; Head requests should be treated exactly like GET requests. In previous years, a HEAD request bypassed the J2EE servlet method restrictions, so unauthenticated users could potentially access functionality that did not check the method properly. So an attacker could do an authenticated request as effectively a pseudo GET and all the parameters were in the URL. 

I also test for RIDICULOUS, to see what happens when the application framework gets an unexpected method.

**Comment by kingthorin** [comment]
I don't think anyone suggested OPTIONS was an issue, if they did probably only as an info leak, though thats a stretch as it fulfills a specific purpose.
PUT and DELETE aren't WebDav specific they're part of the base spec https://www.w3.org/Protocols/rfc2616/rfc2616-sec9.html, though I agree could represent some danger for automated testing. Hence the earlier suggestion of tying to specific scan strengths.
List of WebDav methods here: https://www.w3.org/Protocols/rfc2616/rfc2616-sec9.html
I agree with the idea of implementing separately.

As mentioned back in June 2015 this did get worked on: https://code.google.com/p/zap-extensions/source/detail?r=1351
So far handles TRACE, TRACK,  and CONNECT.

**Comment by thc202** [comment]
The library does allow to send non-standard methods.

**Comment by 3bhady** [comment]
I'd like to tackle this issue if no one is on it.

**Comment by kingthorin** [comment]
You could look at this abandon PR and put something together: https://github.com/zaproxy/zap-extensions/pull/837

**Comment by 3bhady** [comment]
Thanks @kingthorin  , I'm on it.

**Comment by 3bhady** [comment]
@kingthorin,  He has already removed the fork
Should I start my own fork and use his code as guiding line?

**Comment by kingthorin** [comment]
Yup. Just read through the discussion in the PR (so that you avoid any of the outstanding issues). Create your own local feature branch and implement some changes. Then create your own PR. 

We will just close that original one when the new one is being reviewed or something...

**Comment by 3bhady** [comment]
@kingthorin this is my first issue so excuse the time I've spent messing around to get everything working,
I've managed to get the old PR running and tested it a little bit,
what I've seen so far is that testing the methods is working fine except for PUT which is not handled with a file as a payload.
so what is needed to be done is 
- handle files as a payload in PUT requests 
is that correct or am i missing something?

**Comment by kingthorin** [comment]
That was one of the outstanding issues, I think some of the alert information might need to also be updated to reflect the realities of various verbs (methods) being used by web APIs.

**Comment by 3bhady** [comment]
@kingthorin I've fixed PUT and PATCH requests,
but I was wondering should i just send a `"dummyKey":"dummyVal"` as body message or should i try something like php code and see how the server would respond?

About the alerts can you tell me what kind of alerts that are wrong? like the description of the alert or the evidence?

Now should i create a pull request after finishing everything or should i finish fixing the alerts first?

Also after finishing the above I was wondering about the [429 too many requests] response should we do something about this? like lower the rate of scanning or something?

**Comment by kingthorin** [comment]
I'd go with safe values, and a randomly generated filename (so that nothing is accidentally clobbered or linked on the target/server). 

I haven't looked at the alert content to see what might need to be changed. I was simply mentioning it in relation to this comment: https://github.com/zaproxy/zap-extensions/pull/837#issuecomment-372738299

I'd suggest that you read through the content of the alert(s) and see if there's anything that doesn't make sense if you're considering modern use of Web APIs vs. the original intention of the verbs (as defined in RFC).

I'd say wait until you've done as much as you can. Hopefully including some local testing (you can leverage XAMPP Portable or something like that.)

Have you encountered a 429 while putting this together? I guess you could check for a 429 while looping through the methods and then raise an Informational alert if one is encountered (and exit the loop).

**Comment by 3bhady** [comment]
@kingthorin I've encountered 429 , If the Spider runs for a while but not long enough to get a 429 , Running active scan after it causes 429.

I'll try to finish them as soon as possible so that i can deliver the issue before GSoC.

**Comment by psiinon** [comment]
@3bhady what application are you testing this against? I'd use a local test application so that you are more in control.
I'm not aware of any of our existing rules explicitly handling 429 responses - if a app stops talking to use because we've made too many requests then theres not a great deal we can do.
We do have the global option to sleep between requests, you dont need to do anything to support that as its handled by the core.

**Comment by 3bhady** [comment]
@psiinon I was testing it against public programs in Hackerone like github.com itself.
But i can use DVWA if that would be better.
I was thinking of delaying the requests when getting 429 response and repeating the lost request untill the blocking period has passed, but if it's not handled by any of the existing rules then I think we should leave it like it is.
I'll try to finish the basics and make a Pull Request first.

**Comment by psiinon** [comment]
I wouldnt try to handle this in each rule, it would better for us to handle it in the core for all rules.
Get the basics done and we can worry about 429s later :)

**Comment by 3bhady** [comment]
@psiinon @kingthorin I've submitted a Pull Request, please let me know if there is something wrong so that i can change it.

**Comment by thc202** [comment]
Fixed in zaproxy/zap-extensions#1500.

**Comment by thc202** [comment]
Released in version 24 of Active scanner rules (beta) add-on.
